### ISSUE TITLE
Allow type guards in getters

### ISSUE THREAD

**Issue by ashmind** [microsoft/TypeScript#52078]
# Suggestion

## 🔍 Search Terms

getters type guards accessors
A type predicate is only allowed in return type position for functions and methods. ts(1228)

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I would like to see type guards allowed on getters, in the same way they are currently allowed on other members. This would make the language more consistent. Currently it feels like choice between getter and a method is sometimes enforced by TypeScript (depending on whether you need a guard or not).

> **Note**
> This is not a duplicate of https://github.com/microsoft/TypeScript/issues/7150. 7150 is a bug report, closed by design, but this is a request to change the design.

## 📃 Motivating Example

From JS perspective both options below are pretty much equivalent, so there is no reason for TS to allow one but not the other:

```typescript
class ExampleWithMethod {
  prerequisite: string | undefined;

  isReady(): this is { prerequisite: string } {
    return this.prerequisite !== undefined;
  }

  doWork() {
    if (!this.isReady())
      throw new Error("Not ready");

    console.log(this.prerequisite.length);
  }
}

class ExampleWithGetter {
  prerequisite: string | undefined;

  get isReady(): this is { prerequisite: string } {
    return this.prerequisite !== undefined;
  }

  doWork() {
    if (!this.isReady)
      throw new Error("Not ready");

    console.log(this.prerequisite.length);
  }
}
```

https://www.typescriptlang.org/play?#code/MYGwhgzhAEAqCmEAuBGaBvAUNaAHATvIQI4CuAlhOUvAFzTL7kB2A5tAD7SnMAm8AMxbxeAbkzZolAErwwvAJ4AKAJT0kAC0pSY6PIRIUqNeoxbsAvhkk5CSUvmbRNlAHQEi8MpWrxoAQgBeQO4+QWExSQsJHF4AewB1OPwAa1VrHBxyAWglfxcIVxk5RVUVG0zNfDiAd2hmeDqAUXxq-CUAIgA5OKRoQnkFDpVxCuA45gg4kHhXEDjWJQL3Ay8jXzn4Nk0RqMxozFBIGARkACYM-U9vYzoGJCY2TlD+IQbIyVZ4PuLB1XUtDBtHoPIYfCZ7o9LJdbN8HE5lqC1uC-EEQjxXhFxDgDrFEsk0ioYVIcnllr9FOVMpUNNU6g1mq1kp0en0BophqNqeNJtNZvNFojVjcNjNthpdjj9kA

## 💻 Use Cases

As per example, I would like to use this to conditionally narrow `this` within more complex classes. It can also be used externally outside of the class (e.g. https://github.com/microsoft/TypeScript/issues/7150#issuecomment-187182400), though it is likely that most external cases can be solved through unions.

It's hardly a common case, but it does feel like an unexpected inconsistency within language.

**Comment by jcalz** [comment]
Duplicate of #43368
